### PR TITLE
[Reflection] Implement an SPI to iterate types who conform to some protocol

### PR DESCRIPTION
**Pull request by Azoy**
Resolves: rdar://105290756

### CODE REVIEW

**Comment by Azoy** [comment]
@swift-ci please smoke test

File: stdlib/public/Reflection/Sources/Reflection/Conformances.swift [hunk 1]
```diff
@@ -0,0 +1,70 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
```
Comment (mikeash) [inline stdlib/public/Reflection/Sources/Reflection/Conformances.swift:5]: Not terribly important, but it's now 2023.

File: stdlib/public/Reflection/Sources/Reflection/Conformances.swift [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  withUnsafeMutablePointer(to: &result) {
+    _withConformanceCache(proto, UnsafeMutableRawPointer($0)) {
+      let buffer = UnsafeBufferPointer<Any.Type>(
+        start: UnsafePointer<Any.Type>($0._rawValue),
+        count: $1
+      )
+
+      let arrayPtr = $2.assumingMemoryBound(to: [Any.Type].self)
+
+      arrayPtr.pointee.reserveCapacity($1)
```
Comment (mikeash) [inline stdlib/public/Reflection/Sources/Reflection/Conformances.swift:59]: This copy loop can be replaced by `arrayPtr.pointee = Array(buffer)` which would be a bit nicer and perhaps faster.

**Review by mikeash — APPROVED** [review]
Looks good. /cc @grynspan

**Comment by Azoy** [comment]
@swift-ci please smoke test

**Comment by grynspan** [comment]
This PR includes a lot of code duplicated from the main runtime sources. I don't think that's the right approach here. Surely we can just bubble something up from the runtime's module directly?

**Comment by Azoy** [comment]
@mikeash @grynspan I've moved some of the common definitions from the runtime into the include header so the reflection module can easily use it. Let me know what you think.

**Comment by grynspan** [comment]
Why can't you just include the headers at their old locations?

**Pull request closed without merging**